### PR TITLE
Add subresource for manually instantiating CronJobs

### LINKED ISSUE #47538: Ability to start CronJobs manually
It would be nice to support manual runs for CronJobs. Today, IIRC, a CronJob misses its schedule for whatever reason and we won't start it until the next time it's scheduled to run. At the very least, this could be a kubectl generator, eventually we may want to support it as part of the CJ API.

@stevekuznetsov since you were asking about it today

@kubernetes/sig-apps-feature-requests

### PR DESCRIPTION
**Pull request by erhudy**
**What this PR does / why we need it**: makes it possible to manually instantiate CronJobs for debugging/testing purposes

**Which issue this PR fixes**: fixes #47538

**Release note**:
```
CronJobs now have an /instantiate subresource that, when POSTed to, will immediately spawn a Job from the spec in the CronJob.
```

### CODE REVIEW

**Comment by erhudy** [comment]
This PR still has some outstanding issues. I am putting it up in advance of today's sig-apps discussion so that it is readily available for comment and review.

File: pkg/apis/batch/types.go [hunk 1]
```diff
@@ -238,6 +238,20 @@ type CronJobList struct {
 	Items []CronJob
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CronJobManualInstantiation represents a manual triggering of a CronJob.
+type CronJobManualInstantiation struct {
```
Comment (soltysh) [inline pkg/apis/batch/types.go]: I was thinking this will be called `CronJobManualRequest` or `CronJobRequest`, for example, wdyt?
@kubernetes/api-reviewers thoughts?

File: pkg/apis/batch/types.go [hunk 2]
```diff
@@ -238,6 +238,20 @@ type CronJobList struct {
...
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// CreatedJob is the Job created in the instantiation request.
+	CreatedJob Job
```
Comment (soltysh) [inline pkg/apis/batch/types.go]: This is not needed, I'd expect the `/instantiate` endpoint to return you that `Job` instance.

File: pkg/registry/batch/cronjob/storage/storage.go [hunk 3]
```diff
@@ -84,3 +101,53 @@ func (r *StatusREST) Get(ctx genericapirequest.Context, name string, options *me
...
+func (r *InstantiateREST) New() runtime.Object {
+	return &batch.CronJobManualInstantiation{}
+}
+
+func (r *InstantiateREST) Create(ctx genericapirequest.Context, name string, obj runtime.Object, includeUninitialized bool) (out runtime.Object, err error) {
+	sourceCronJobObj, err := r.cronJobStore.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	sourceCronJob, ok := sourceCronJobObj.(*batch.CronJob)
```
Comment (soltysh) [inline pkg/registry/batch/cronjob/storage/storage.go]: I was hoping that `Create` will accept a `CronJobRequest` object and find the appropriate `CronJob` and return the actual `Job` as a result. See OpenShift's [build/instantiate](https://github.com/openshift/origin/blob/9f48098688b100e4d5f7b889818eab826eedb774/pkg/build/registry/buildconfiginstantiate/rest.go#L60-L75).

File: pkg/kubectl/cmd/create_job.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/batch/internalversion"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/util/i18n"
+)
+
+var (
+	jobLong = templates.LongDesc(i18n.T(`
+		Create a Job to execute immediately from a specified CronJob.`))
```
Comment (soltysh) [inline pkg/kubectl/cmd/create_job.go]: Since you're creating `kubectl create job`, which is fine, because I was also thinking about doing it, you need to:

1. make sure it creates just a job, by default
2. only then adding `--from-cronjob` flag.

Comment (soltysh) [reply pkg/kubectl/cmd/create_job.go]: For one see how `kubectl create deployment` works, using a generator and do something similar, and should limit the amount of code needed to write it.

File: staging/src/k8s.io/client-go/Godeps/Godeps.json [hunk 5]
```diff
@@ -1,16 +1,12 @@
 {
 	"ImportPath": "k8s.io/client-go",
-	"GoVersion": "go1.9",
+	"GoVersion": "go1.8",
```
Comment (soltysh) [inline staging/src/k8s.io/client-go/Godeps/Godeps.json]: I doubt you should be touching those files.

File: pkg/registry/batch/cronjob/storage/storage.go [hunk 6]
```diff
@@ -124,10 +124,17 @@ func (r *InstantiateREST) Create(ctx genericapirequest.Context, name string, obj
 		return nil, fmt.Errorf("got object of type %T, expected *batch.CronJob", sourceCronJobObj)
 	}
 
+	// add an extra label onto the created job to signal to the cronjob controller that it should adopt this job
+	jobLabels := make(map[string]string)
+	for _, key := range sourceCronJob.Spec.JobTemplate.Labels {
+		jobLabels[key] = sourceCronJob.Spec.JobTemplate.Labels[key]
+	}
+	jobLabels["createdByInstantiate"] = "yes"
```
Comment (soltysh) [inline pkg/registry/batch/cronjob/storage/storage.go]: Annotation will be better for this type of information, name it something like `cronjob.kubernetes.io/instantiate=manual`

Comment (soltysh) [reply pkg/registry/batch/cronjob/storage/storage.go]: @kubernetes/api-reviewers thoughts?

File: pkg/registry/batch/cronjob/storage/storage.go [hunk 7]
```diff
@@ -84,3 +101,53 @@ func (r *StatusREST) Get(ctx genericapirequest.Context, name string, options *me
...
+		return nil, err
+	}
+	createdJob, ok := createdJobObj.(*batch.Job)
+	if !ok {
+		return nil, fmt.Errorf("got object of type %T, expected *batch.Job", createdJobObj)
+	}
+
+	// TODO update the CronJob so that it knows about the instantiated job
+	// not currently done because clobbering the CronJob in the store would create a race condition that
+	// could confuse the CronJob controller, which expects to have exclusive control over a CronJob's status
```
Comment (soltysh) [inline pkg/registry/batch/cronjob/storage/storage.go]: I think this TODO is addressed in the 3rd commit so it can be removed from here.

File: pkg/controller/cronjob/cronjob_controller.go [hunk 8]
```diff
@@ -210,12 +210,24 @@ func syncOne(sj *batchv1beta1.CronJob, js []batchv1.Job, now time.Time, jc jobCo
...
+
+			// if the job was created via the instantiate subresource, adopt it
+			if createdByInstantiate, exists := j.Labels["createdByInstantiate"]; exists && createdByInstantiate == "yes" {
+				ref, err := getRef(&j)
+				if err != nil {
+					glog.V(2).Infof("Unable to make object reference for job for %s", nameForLog)
+				}
+				sj.Status.Active = append(sj.Status.Active, *ref)
+
+				// should we update LastScheduleTime, since this job was not scheduled?
```
Comment (soltysh) [inline pkg/controller/cronjob/cronjob_controller.go]: No, the `LastScheduleTime` is only for automatic triggering and manually triggering should not interfere with previously running. Not sure how this applies to concurrency policy, I think it will interfere with it, but not for the time of next scheduling.

Comment (soltysh) [reply pkg/controller/cronjob/cronjob_controller.go]: Add a comment explaining this here, please.

**Review by soltysh — CHANGES_REQUESTED** [review]

**Comment by soltysh** [comment]
@erhudy aside from that here are some general commit organization suggestions:

- split the first commit into two (manual changes + generated), it's easier to review, and make them API changes only
- 3rd commit should be the storage changes
- some of the changes from you 'Implement CLI interface' should be split into generated ones (client) and others into either storage or actual separate cli changes
- 4th should be incorporated into one of those groups

File: staging/src/k8s.io/client-go/Godeps/Godeps.json (hunk 5 above)
Comment (erhudy) [reply staging/src/k8s.io/client-go/Godeps/Godeps.json]: I think this got pulled in accidentally by a revase.

File: pkg/apis/batch/types.go (hunk 2 above)
Comment (erhudy) [reply pkg/apis/batch/types.go]: I did it this way because I could not figure out how to circumvent the API machinery in such a way to allow me to return a different type of object than the one specified in the subresource's resource.

Generally speaking, I could not find any other examples of doing what I did, which is to have a subresource return a completely different type of object from its resource. I mostly referenced `pod/logs` when writing this code, which does not return a `Pod` instance.

File: pkg/kubectl/cmd/create_job.go (hunk 4 above)
Comment (erhudy) [reply pkg/kubectl/cmd/create_job.go]: I didn't implement a generic `create job` command because I considered it to be out of scope. I'm happy to do so, but maybe it should be part of a separate PR to avoid blowing the size of this one up even more?

File: pkg/registry/batch/cronjob/storage/storage.go (hunk 3 above)
Comment (erhudy) [reply pkg/registry/batch/cronjob/storage/storage.go]: As in, you think the subresource should operate on a completely different type of object? I felt that doing so was inelegant and would result in fighting against the API machinery, which I don't have enough knowledge about Kubernetes to know how to do. The NamedCreater interface made it straightforward to look up the source CronJob.

**Comment by janetkuo** [comment]
This needs some e2e tests.

**Comment by erhudy** [comment]
Waiting for https://github.com/kubernetes/kubernetes/pull/56003 to get merged and then I will pick up development work again.

File: pkg/apis/batch/types.go (hunk 1 above)
Comment (soltysh) [reply pkg/apis/batch/types.go]: I spoke with @smarterclayton the suggestion is either `CronJobManualRequest` or `CronJobInstantiate`.

File: pkg/apis/batch/types.go (hunk 2 above)
Comment (soltysh) [reply pkg/apis/batch/types.go]: See OpenShift's [build/instantiate](https://github.com/openshift/origin/blob/9f48098688b100e4d5f7b889818eab826eedb774/pkg/build/registry/buildconfiginstantiate/rest.go#L60-L75). The api machinery is not restrictive in this way. You should not have any problems with returning a different object.

File: pkg/registry/batch/cronjob/storage/storage.go (hunk 3 above)
Comment (soltysh) [reply pkg/registry/batch/cronjob/storage/storage.go]: The problem with this is that you need to pass the entire cronjob object just to instantiate it. Besides API server already knows the object. Besides if you're creating a Job it's better to create it from the CronJob that is on the server, not from the one you got, this might lead to differences you don't want to handle here.

File: pkg/registry/batch/cronjob/storage/storage_test.go [hunk 9]
```diff
@@ -184,4 +185,36 @@ func TestWatch(t *testing.T) {
 	)
 }
 
+func TestInstantiate(t *testing.T) {
+	// scheduled jobs should be tested only when batch/v2alpha1 is enabled
+	if *testapi.Batch.GroupVersion() != v2alpha1.SchemeGroupVersion {
```
Comment (soltysh) [inline pkg/registry/batch/cronjob/storage/storage_test.go]: I care more about `batch/v1beta1` than alpha, can we test that instead or both?

File: pkg/registry/batch/cronjob/storage/storage.go [hunk 10]
```diff
@@ -84,3 +101,49 @@ func (r *StatusREST) Get(ctx genericapirequest.Context, name string, options *me
...
+	sourceCronJobObj, err := r.cronJobStore.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	sourceCronJob, ok := sourceCronJobObj.(*batch.CronJob)
+	if !ok {
+		return nil, fmt.Errorf("got object of type %T, expected *batch.CronJob", sourceCronJobObj)
+	}
+
+	job := &batch.Job{
```
Comment (soltysh) [inline pkg/registry/batch/cronjob/storage/storage.go]: Make [getJobFromTemplate](https://github.com/kubernetes/kubernetes/blob/2cbb07a4394ef4cfc1666e114674a047023ed178/pkg/controller/cronjob/utils.go#L173-L194) public and move it here, so that storage (which is part of API server) does not depend on controller, that's not good.

File: pkg/kubectl/cmd/create_job.go (hunk 4 above)
Comment (soltysh) [reply pkg/kubectl/cmd/create_job.go]: Separate PR sgtm.

**Review by soltysh — CHANGES_REQUESTED** [review]
The previous comments still apply, as well and a few others.

File: pkg/registry/batch/cronjob/storage/storage.go (hunk 6 above)
Comment (soltysh) [reply pkg/registry/batch/cronjob/storage/storage.go]: Comment still applies.

**Comment by janetkuo** [comment]
Does this PR target 1.9? Code slush is today and code freeze is approaching (Wed).

**Comment by erhudy** [comment]
It will take me some time to address @soltysh's requested changes around my normal work responsibilities. As far as I'm aware, it was never committed to for 1.9.

**Comment by janetkuo** [comment]
@erhudy understood, thanks! I'm asking because this feature was mentioned in https://github.com/kubernetes/features/issues/19#issuecomment-340060472

File: pkg/apis/batch/types.go (hunk 2 above)
Comment (erhudy) [reply pkg/apis/batch/types.go]: Here's the problem I'm having with accommodating this request. It wasn't a problem with the type of the object, but rather the version.

I rewrote the genclient line as:
```
// +genclient:method=Instantiate,verb=create,subresource=instantiate,input=k8s.io/api/batch/v1beta1.CronJobManualRequest,result=k8s.io/api/batch/v1.Job
```
...which generated:
```
// Instantiate takes the representation of a cronJobManualRequest and creates it.  Returns the server's representation of the job, and an error, if there is any.
func (c *cronJobs) Instantiate(cronJobName string, cronJobManualRequest *v1beta1.CronJobManualRequest) (result *batch_v1.Job, err error) {
    result = &batch_v1.Job{}
    err = c.client.Post().
        Namespace(c.ns).
        Resource("cronjobs").
        Name(cronJobName).
        SubResource("instantiate").
        Body(cronJobManualRequest).
        Do().
        Into(result)
    return
}
```
...and that seems all fine.

However, hitting the API server seems to fail at the `Into(result)` point with:
```
{
    "kind": "Status",
    "apiVersion": "v1",
    "metadata": {},
    "status": "Failure",
    "message": "no kind \"Job\" is registered for version \"batch/v1beta1\"",
    "code": 500
}
```

It doesn't seem to be able to mix API versions. The request works up to this point; the job is created as expected, but `Into()` seems to be trying to materialize a Job object of kind `batch/v1beta1`.

**Comment by soltysh** [comment]
> @erhudy understood, thanks! I'm asking because this feature was mentioned in kubernetes/features#19 (comment)

We were hoping but never committed to anything.

**Comment by erhudy** [comment]
I brought the above issue up in the API machinery SIG Slack room - @liggitt and @lavalamp both responded to say that wanting to cross versions was highly unusual. Given that, I would like to either wait for CronJob to be promoted to batch/v1, or keep the current setup where the CJMR kind is used as an envelope for the created job. (It does not require any knowledge of the existing job in order to create it, you just hit the `/instantiate` subresource with minimal object metadata and it takes the appropriate action.)

File: pkg/controller/cronjob/cronjob_controller.go [hunk 11]
```diff
@@ -215,12 +215,23 @@ func syncOne(sj *batchv1beta1.CronJob, js []batchv1.Job, now time.Time, jc jobCo
...
-			// or if our jobs list is newer than our sj status after a relist, or if someone intentionally created
-			// a job that they wanted us to adopt.
+			// This could happen if:
+			// * we crashed right after creating the Job and before updating the status
+			// * if our jobs list is newer than our sj status after a relist
+			// * if someone intentionally created a job that they wanted us to adopt
+			// * if a job was created via the instantiate subresource (will be labeled as createdByInstantiate=yes)
+
+			// if the job was created via the instantiate subresource, adopt it
+			if instantiateAnnotation, exists := j.Annotations["cronjob.kubernetes.io/instantiate"]; exists && instantiateAnnotation == "manual" {
```
Comment (soltysh) [inline pkg/controller/cronjob/cronjob_controller.go]: Can you make the above warning different for manual jobs, it should along the lines of controller noticed manual job...

File: pkg/kubectl/cmd/create_job.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (soltysh) [inline pkg/kubectl/cmd/create_job.go]: 2018

File: pkg/apis/batch/types.go [hunk 13]
```diff
@@ -238,6 +239,20 @@ type CronJobList struct {
 	Items []CronJob
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CronJobManualRequest represents a manual triggering of a CronJob.
+type CronJobManualRequest struct {
```
Comment (soltysh) [inline pkg/apis/batch/types.go]: I'm still struggling with two options atm:

1. Take your current options which embeds the created job.
2. Just return `metav1.Status` from that endpoint.

**Review by soltysh — CHANGES_REQUESTED** [review]
I left you a few comments. I'll try to figure out by the EOD which approach (embeded job or status) you should go with. And I'd like to push it over the hump by EOW.

File: pkg/apis/batch/types.go (hunk 13 above)
Comment (soltysh) [reply pkg/apis/batch/types.go]: @kubernetes/api-reviewers any preference?

**Comment by deads2k** [comment]
There are existent examples of instantiate endpoints in the openshift API, but if you look at something like `buildconfig/instantiate` the podtemplate being created isn't available to the user and the user doesn't have sufficient PSP rights to produce the pod.  The build API provides a constrained escalation path using known safe podtemplates.

The cronjobs/instantiate looks different to me.  The cronjob includes a jobtemplatespec inside of it and the expectation is that the user can create the job in question.  Rather than risk a delegated authority problem (user calls the apiserver under a cronjob which then makes another call on the user's behalf to create a job), why not simply tell the user to pull the jobtemplatespec from the cronjob and create it directly?  The forces driving the choice are different, since the permissions for a cronjob/job behave differently than the downstream API.

/cc @kubernetes/api-reviewers @kubernetes/api-approvers

**Comment by bgrant0607** [comment]
This would set a precedent for the API, so we need to be somewhat careful.

The discussion is a bit hard to follow. Is this launching a job as if periodically triggered, or just returning a Job that the client could POST?

cc @kow3ns @erictune

**Comment by erhudy** [comment]
My only goal was to provide a convenience mechanism for immediately invoking an existing CronJob via `kubectl` instead of having to wait for the next scheduled run. I am entirely fine with dropping everything here and rewriting it all to just be a bit of logic in `kubectl` to fetch a CronJob, extract the JobSpec, then turn it into a new Job.

**Comment by soltysh** [comment]
> The discussion is a bit hard to follow. Is this launching a job as if periodically triggered, or just returning a Job that the client could POST?

The latter, that's why the simplest approach here is to just move that bit to client, who can take the jobTemplate from CronJob and post it to server, just like the controller normally does.

**Comment by soltysh** [comment]
I'll try to bring up this topic during tomorrow's sig-cli and for Thursday sig-arch call as well.

**Comment by erhudy** [comment]
Based on my discussions with @soltysh, it does not sound like the community is willing or able to agree on an approach here. I just wanted to implement a convenience mechanism for kicking a CronJob off prior to its scheduled runtime, but this has been bogged down for 5 months with no end in sight.

I am abandoning this PR and closing it.

**Comment by berlincount** [comment]
:(
